### ISSUE TITLE
real time super slow

### ISSUE THREAD

**Issue by tiralonghipol** [ethz-asl/rovio#192]
Hi. I calibrated my camera with Kalibr, and setup everything.
I was wondering which parameters can I act on in order to speed up the performance for real time..
Right now, seems like rovio keeps doing calculations on the first frame on startup and if I move the camera the frame still fixed.
I'm using an i7 pc, and the cpu in not saturated..
any hint?
Thank you!

**Comment by helenol** [comment]
Rovio should run real-time at about 20-30 Hz using about ~20% CPU on a standard laptop CPU with default parameters out of the box. Make sure you compile in Release mode (and not debug) and that the data is piped in correctly, and that you're feeding in IMU data that's time synchronized to the camera. Rovio will not work without IMU data.

**Comment by tiralonghipol** [comment]
Ok then I'm assuming some bad camera/imu sync.. how to check that? any suggestion?
thank you for your reply, btw

**Comment by helenol** [comment]
Maybe the easiest thing to do is to first check that it's running correctly on the euroc dataset, and in real-time. This will rule out that it's anything other than your data.
Probably the easiest way to check your camera-IMU time sync is to run kalibr ( https://github.com/ethz-asl/kalibr ) for cam-IMU calibration. If you're getting huge reprojection errors, or the motions aren't matching, then this is your issue.

**Comment by ZacharyTaylor** [comment]
The output timeshift from kalibr is currently 0.0s. This indicates that it was run without time calibration. Could you rerun it with the `--time-calibration` flag as this will let us know if there is any offset in timestamping between your camera and IMU

**Comment by tiralonghipol** [comment]
I feel really stupid.. thank you!
What if the seller declared "hardware camera-imu sync" ?
Does the `--time-calibration` flag still valid ?

**Comment by ZacharyTaylor** [comment]
If the system has accurate hardware timestamping than `--time-calibration` will output something like 1ms or less of offset verifying that the timestamping is accurate.

That error occurs as the system has a larger time offset than kalibr is expecting (default max of 0.01s). Try setting `--timeoffset-padding 0.1` and rerunning. However this already suggest the hardware timestamping has some offset in it. Out of curiosity which sensor are you using?

**Comment by tiralonghipol** [comment]
Here is the link to the sensor:
https://www.robotshop.com/en/optor-visual-inertial-camera.html 
Today I'll give a try to your last suggestion, thank you!

**Comment by ZacharyTaylor** [comment]
So I have bad news, if this sensor is doing hardware triggering / timestamping than the driver is throwing this information away and just stamping things on arrival in software. 

Here is the line in the driver it happens at https://github.com/optor-vis/vi-stereo-v1/blob/master/optor_VI_Sensor_SDK_V1.0/SDK/src/optorcam.cpp#L954 

The two cameras have different timing offsets because they are sent one at a time and so have different arrival times. While an offset of 1 or 2ms probably wouldn't be a deal breaker with 16ms of offset things will probably diverge pretty much immediately

**Comment by tiralonghipol** [comment]
Thanks again for your support and your patience!
So now what do you suggest? I think the options are 2: modify the driver o use other sensors, am I right?

**Comment by ZacharyTaylor** [comment]
If no hardware timestamps exist you don't have that many good options. You can take the values kalibr gives you and code these into the driver to get things a bit more consistent. However, these offsets will change with factors such as your CPU load and the arrival times will usually have a spread of a few milliseconds. 

![time](https://raw.githubusercontent.com/wiki/ethz-asl/cuckoo_time_translator/images/chameleon10kHz-with-legend.png)

Since you know the images and IMU are running at a fixed rate you can implement a simple kalman filter to smooth out this ripple and try hardcode the offset, though I am unsure how accurate this will get you.

Most people using VI sensing have their own custom solution. Of the commercial solutions I know of that we have tested:
The S360 sensor - https://zurichsense.com/ expensive and made in low numbers, but we have had good results with it
The DUO - https://duo3d.com/ a little pricey and from memory the IMU is heavily filtered. We had one working but not reliably.
Realsense ZR300 - out of production, but maybe you can find one. A little sensitive to shocks and we had to modify the provided driver, but it does work.
ZED Mini - the maplab guys just got one of these. So far it seems like it works though we are still testing

**Comment by tiralonghipol** [comment]
many thanks again!
Another question (a little off topic): have you ever tried rovio in dark or completely dark environments? or what could be the best camera/sensor for low light condition?

**Comment by ZacharyTaylor** [comment]
One with a torch attached. Though seriously we have managed to run rovio on images that are so dark a person can barely see whats going on. As a rule of thumb the bigger the sensor and the larger the lens the more light will be captured, however you hit diminishing returns pretty quickly making active illumination a good option for dark enviroments.

**Comment by tiralonghipol** [comment]
Do you have any rosbag to share? or related video/paper about these experiments?

**Comment by ZacharyTaylor** [comment]
The darkest flights we have done were probably in the back of the building shown here https://www.youtube.com/watch?v=49WiPHkkhBw

Here is a short section of rovio running on part of one of these flights

![peek 2018-08-02 16-54](https://user-images.githubusercontent.com/730680/43591988-dfdd8c16-9674-11e8-82c5-5db8d1d8f729.gif)

**Comment by tiralonghipol** [comment]
This is reeeeally impressive! and also what I was looking for! 
Thanks again

**Comment by Razbotics** [comment]
What is the status of rovio running with optor camera? I have a optor camera and i have finished with calibration... but i get slow odometry results

**Comment by ZacharyTaylor** [comment]
If you are getting slow odometry this usually means that rovio has diverged and it is falling behind real time while it tries to find new features in a large region.

Unfortunately the cause of this will be that tho optor camera relies on poor time synchronization through arrival time timestamping. This combined with the fairly low quality IMU means it is unlikely to ever work well with rovio.

**Comment by Razbotics** [comment]
Hey thanks for the info...though its a bad news for me...well is there any workarounds for better timestamping?....i found that VINS mono is working quite well with optor cam and i heard that rovio works more robust than VINS mono using optor cam..

**Comment by Razbotics** [comment]
Is there a way to hack trigger pins of optor camera? I want them for hardware synchronization

**Comment by ZacharyTaylor** [comment]
I believe VINS has online time offset estimation, which allows it to compensate for these sort of issues. I am unsure about what is possible with the optor. I have never used one, just looked through its driver.

**Comment by Razbotics** [comment]
Yep ViNS mono can compensate that issue and maybe thats the reason VINS works quite well. Do you have knowledge of hardware pins of a camera module? I can see some accessible pins HS, VS , LED out and PCLK on the back of optor cam. I was hoping if i can use one of them for hardware timestamping and synchronisation.

**Comment by chenwinki** [comment]
@tiralonghipol  Hello, could you give me a hint how to get the param of imu like this ,thank you very much:
Accelerometer:
Noise density: 0.005
Noise density (discrete): 0.0678232998313
Random walk: 0.00433
Gyroscope:
Noise density: 0.005
Noise density (discrete): 0.0678232998313
Random walk: 0.00266

**Comment by tiralonghipol** [comment]
@chenwinki  can you be more specific on your question please? what do you mean by 'get the param of imu'
If you are talking about the configuration file (-yaml), you do not get anything, you set it as you like (check che datasheet of your IMU, of course)
hope it helps

**Comment by chenwinki** [comment]
@tiralonghipol Thank you for you advice! I calibrate imu using imu_utils, and calibrate imu and camera using kalibr. but I got bad result, even when I put my VI sensor static, the result shows the VI sensor are moving(about ten meter  in 100s), could you give me a hint about the result?

**Comment by tiralonghipol** [comment]
The Optor / Loitor VI sensor has many problems, being a cheap sensor. The worst one is that the sync between images and imu is not really hardware
After many tries with different algorithms I decided to switch to the MyntEye camera and I immediately got good results (after proper calibration, of course)
For more details see the discussion above https://github.com/ethz-asl/rovio/issues/192#issuecomment-408401898

**Comment by mechaliomar** [comment]
@tiralonghipol  So, you have been using the mynteye camera successfully!
-I am also planning to use it for 6 DOF pose estimation using ROVIO.
-Can you suggest something before I start ? 

BEST REGARDS
UESTC, CHINA

**Comment by tiralonghipol** [comment]
Hi @mechaliomar 
I ended up in building my own visual inertial sensor. Unfortunately the commercial sensors I tested during this year were not as good as declared
Currently i'm using rovio onboard an autonomous drone and the results are really good
https://youtu.be/N7SK1ekzhe0

**Comment by arashazimi0032** [comment]
hi all, i was runing ROVIO on EuRoC data set with i7 cpu and 8 gb RAM on ubuntu 18.04 but it is very very slow and It diverges after a short time. what is the problem?
Of course, I build it with catkin_make not catkin build Because catkin build was giving an error about the lightweight_filtering. please guide me. Tanks
